File: discord-demibot/src/http/routes/me.js
const express = require('express');

module.exports = ({ db, discord }) => {
  const router = express.Router();
  const client = discord.getClient();

  router.get('/', (req, res) => {
    const info = req.apiKey || {};
    res.json({ userId: info.userId, isAdmin: !!info.isAdmin });
  });

  router.get('/roles', async (req, res) => {
    const { userId } = req.query;
    const info = req.apiKey;
    if (!userId || userId !== info.userId) {
      return res.status(403).json({ hasOfficerRole: false });
    }
    try {
      const guild = client.guilds.cache.get(info.serverId);
      if (!guild) {
        return res.status(500).json({ hasOfficerRole: false });
      }
      const member = await guild.members.fetch(userId);
      const roles = Array.from(member.roles.cache.keys());
      const officerRoles = await db.getOfficerRoles(info.serverId);
      const hasOfficerRole = roles.some(r => officerRoles.includes(r));
      res.json({ hasOfficerRole });
    } catch (err) {
      res.status(500).json({ hasOfficerRole: false });
    }
  });

  return router;
};

File: python_demibot/routes/me.py
from fastapi import APIRouter, Depends, HTTPException

from ..api import get_api_key_info, db, bot

router = APIRouter(prefix="/api/me")


@router.get("/")
async def get_me(info: dict = Depends(get_api_key_info)):
    return {"userId": info.get("userId"), "isAdmin": bool(info.get("isAdmin"))}


@router.get("/roles")
async def get_roles(userId: str, info: dict = Depends(get_api_key_info)):
    if userId != info.get("userId"):
        raise HTTPException(status_code=403, detail={"hasOfficerRole": False})
    guild = bot.get_guild(int(info["serverId"]))
    if not guild:
        raise HTTPException(status_code=500, detail={"hasOfficerRole": False})
    try:
        member = await guild.fetch_member(int(userId))
        roles = [str(r.id) for r in member.roles]
        officer_roles = await db.get_officer_roles(info["serverId"])
        has = any(r in officer_roles for r in roles)
        return {"hasOfficerRole": has}
    except Exception as err:  # pragma: no cover - defensive
        raise HTTPException(status_code=500, detail={"hasOfficerRole": False}) from err
